### ISSUE TITLE
Use of variadic functions like printf are incompatible with SYCL kernels

### ISSUE THREAD

**Issue by felker** [IAS-Astrophysics/athenak#582]
Unable to compile current `main` branch on ALCF Aurora with SYCL backend of Kokkos due to the presence of `printf()` calls within Kokkos inline functions and lambdas. The C function is varaidic, which is unsupported in SYCL. The SYCL `stream` class covers this functionality. 

Current compilation error:

```
./src/eos/primitive_solver_hyd.hpp:243:80: error: SYCL kernel cannot call a variadic function
  243 |         printf("Error occurred in PrimToCons at (%d, %d, %d, %d)\n", m, k, j, i);
```
Was able to compile with default pgen and options with Kokkos 4.4.00 (note, there is an error with 4.1.00) after commenting out all `printf` calls within that single file:
- L243
- L448
- L452
- L590
- L594
- L598
- L602
- L606
- L623

Likely also an issue in other Z4c files and e.g. `./pgen/dyngr_tov.cpp` and `./outputs/io_wrapper.cpp`, but I did not test this.

Why are C headers like `#include <stdio.h>` and `math.h`, `float.h` etc. even used instead of the C++ equivalents, or even compatibility headers like `cstdio`?

If there is a good reason for sticking with `printf()`, suggested to switch to `Kokkos:printf`, which was released in 4.2.00 https://kokkos.org/kokkos-core-wiki/API/core/utilities/printf.html

**Comment by jfields7** [comment]
> Why are C headers like #include <stdio.h> and math.h, float.h etc. even used instead of the C++ equivalents, or even compatibility headers like cstdio?

We found a really weird bug early in development with certain math functions that led us to do it this way. I don't know if it's the case anymore, but at one point we were advised against using `std::` inside CUDA kernels, so we stuck with the C variants instead via headers like `cmath`. However, this caused issues with some versions of GCC, which wouldn't always select the correct math functions anymore and caused the torus pgen to generate complete nonsense. The long story short is that headers like `cmath` guarantee that all C functions will be available in the STL namespace, but there's actually no guarantee, despite being C compatibility headers, that they're available in the global namespace. The older C headers, on the other hand, _do_ guarantee that they're available in the global namespace. The proper long-term solution would be to use neither `cmath` nor `math.h` and switch exclusively to using Kokkos math functions inside kernels.

Kokkos hasn't been updated yet because Kokkos 4.2 gave runtime errors on Perlmutter (see [this commit](https://github.com/IAS-Astrophysics/athenak/commit/e8b7f92415f7073f3c96df11ee0b3e8252c90f2f)). It might be worthwhile to check if these errors persist or if newer versions of Kokkos work.

**Comment by felker** [comment]
That does sound weird! I always thought that CUDA basically strips the namespace and overloads any basic `cmath` function with a GPU counterpart, see https://stackoverflow.com/questions/68058821/why-does-stdsin-work-in-the-cuda-kernel?noredirect=1&lq=1


>  The long story short is that headers like `cmath` guarantee that all C functions will be available in the STL namespace, but there's actually no guarantee, despite being C compatibility headers, that they're available in the global namespace.

Why would you want these functions in global namespace, assuming the above quirk is no longer true?

Might be worth bumping from Kokkos 4.1.00 to 4.5.00 next, whenever they release it. SYCL support will be moved out of the Experimental namespace in that release. No release candidate [branch](https://github.com/kokkos/kokkos/branches/all) yet, but mentioned as an upcoming release here: https://github.com/kokkos/kokkos/issues/7183

**Comment by jfields7** [comment]
@felker I've updated all the `printf` calls in AthenaK to use `Kokkos::printf` instead (see #594). Can you check if the `kokkos-printf` branch works on Intel?

**Comment by felker** [comment]
@jfields7 looks good to me!

**Comment by felker** [comment]
Closed in #594
